### ISSUE TITLE
Extend jekyll serve's --unpublished flag to also filter collections

### ISSUE THREAD

**Issue by philipbelesky** [jekyll/jekyll#7026]
- [x] I updated to the latest Jekyll (or) if on GitHub Pages to the latest `github-pages`
- [x] I ran `jekyll doctor` to check my configuration
- [x] I read the CONTRIBUTION file at https://jekyllrb.com/docs/contributing/
- [x] This is a feature request.

---

- [x] I am on (or have tested on) ***macOS*** 10+

---

## My Reproduction Steps

- Have a Jekyll collection
- Set `published:  false` in the front-matter of a collection item
- Run `jekyll serve --unpublished`

## The Output I Wanted

At this point unpublished items from the collection will still show on the Jekyll site, whereas unpublished posts will not. This behaviour occurs across multiple of my Jekyll sites. I've also created a fresh install using the quick start and confirmed that it occurs there also.

I'm not sure whether this is a bug or the feature working as designed. If it's the latter I'd like to make the case that intuitively the `--unpublished` flag should apply to both posts and collections given their `published` front-manner is identical and users might expect that commands which discriminate based on `published` status would apply equally to both. I generally use `--unpublished` to work on content that is still in a draft stage but as I use collections more and more not having the ability to easily switch between the 'draft' and 'live' version of the site's collections means having to find work arounds to show/hide content.

I'm not super familiar with Ruby but have dug around the source to try and get a handle for where the parameter is passed and how it is fed through to filter the posts. Before attempting a PR I'd like to check whether or not this is an intended behaviour or if there is some sort of technical hurdle that obstructs changing the behaviour here. In the case that this is a trivial change in the code I would very much anyone taking this on as while I'd like to improve my ruby skills I'm not confident I have the time or ability to do so anytime soon.

Thanks also to everyone who helps develop and maintain Jekyll; I've been using it for about five years now and it's always been exactly what I needed!

**Comment by ashmaroli** [comment]
@philipbelesky Hi, can you please confirm that the behavior you're expecting has changed between versions `3.7.x` and `3.8.x`...

Thanks

**Comment by philipbelesky** [comment]
Hi @ashmaroli. Apologies, I didn't even think to consider whether this might be a regression (the docs refer to just posts so I assumed this was somewhat intentional). Having test it out on `3.7.1` and `3.7.3` collections with `published: false` do in fact show when using `--unpublished`. On both `3.8.0` and `3.8.1` the behaviour I described initially occurs.

**Comment by ashmaroli** [comment]
@philipbelesky Thanks for getting back. This looks like a regression we ought to fix with `v3.8.2`..

Are you interested in submitting a PR for this?
If yes, the offending code is : 
https://github.com/jekyll/jekyll/blob/aed72f6464f675c4426324123ee81e94dab3de43/lib/jekyll/collection.rb#L212-L216

The PR has to have an associated Cucumber feature test.
You can view the *blame* for the method to study the commits that introduced the regression and tests currently in place for that functionality.

Regarding Ruby, the following change should patch this bug:

```ruby
    def read_document(full_path)
      doc = Document.new(full_path, :site => site, :collection => self)
      doc.read
     .unless (doc.data["published"] == false && !site.unpublished)
        docs << doc
      end
    end
```

**Comment by philipbelesky** [comment]
Thanks for identifying the relevant code and patch @ashmaroli. Definitely interested in submitting a PR, I'll brush up on Cucumber and give it a shot!
